### PR TITLE
[Model Runner V2] Support zeroing freshly allocated KV blocks for hybrid + fp8 KVCache

### PR DESCRIPTION
**Pull request by izhuhaoran**
## Purpose

The V2 GPU model runner never zeroed freshly allocated KV cache blocks, unlike the V1 runner. For hybrid gdn+attention models (e.g. Qwen3.5), if use flashinfer(trtllm) attn + fp8 KV cache, FlashInfer quantizes the query and dispatches to the TRTLLM attention kernel, which reads the whole KV page (past the written tokens). 

A newly allocated block still held uninitialized memory, so the kernel produced NaN attention → corrupt logits → an invalid sampled token → `vectorized_gather_kernel` with error `index out of bounds` on the first request.

This PR ports the V1 block-zeroing path to the V2 runner: add `_init_kv_zero_meta` / `_zero_block_ids` and consume `scheduler_output.new_block_ids_to_zero`.

### CODE REVIEW

**Comment by izhuhaoran** [comment]
@WoosukKwon @njhill Could you please review this PR when you have time ?

File: vllm/v1/worker/gpu/model_runner.py [hunk 1]
```diff
@@ -1020,6 +1038,9 @@ def execute_model(
             self.add_requests(scheduler_output)
             self.update_requests(scheduler_output)
             self.block_tables.apply_staged_writes()
+            # Zero freshly allocated blocks to avoid stale NaN/data in attention.
+            if scheduler_output.new_block_ids_to_zero:
+                self._zero_block_ids(scheduler_output.new_block_ids_to_zero)
```
Comment (njhill) [inline vllm/v1/worker/gpu/model_runner.py]: Avoid extra method:

Add `self._kv_block_zeroer: KVBlockZeroer | None = None` in `__init__` method, and here:
```suggestion
            if scheduler_output.new_block_ids_to_zero and self.kv_block_zeroer:
                self.kv_block_zeroer.zero_block_ids(scheduler_output.new_block_ids_to_zero)
```
or better for the above to instead `assert self.kv_block_zeroer is not None` in this case?

Also thinking we could add this to the end of `update_requests()` to keep this method clean

File: vllm/v1/worker/gpu/model_runner.py [hunk 2]
```diff
@@ -446,6 +448,22 @@ def initialize_kv_cache(self, kv_cache_config: KVCacheConfig) -> None:
             self.vllm_config,
         )
         self.kv_connector = get_kv_connector(self.vllm_config, kv_caches_dict)
+        self._kernel_block_sizes = kernel_block_sizes
+
+    def _init_kv_zero_meta(self) -> None:
+        """Build KV-block zeroing metadata; invoked from gpu_worker."""
+        self._kv_block_zeroer = KVBlockZeroer(self.device, is_pin_memory_available())
+        self._kv_block_zeroer.init_meta(
```
Comment (njhill) [inline vllm/v1/worker/gpu/model_runner.py]: Let's not use the underscores to be consistent with other fields in this class.

File: vllm/v1/worker/gpu/model_runner.py [hunk 3]
```diff
@@ -446,6 +448,22 @@ def initialize_kv_cache(self, kv_cache_config: KVCacheConfig) -> None:
             self.vllm_config,
         )
         self.kv_connector = get_kv_connector(self.vllm_config, kv_caches_dict)
+        self._kernel_block_sizes = kernel_block_sizes
```
Comment (njhill) [inline vllm/v1/worker/gpu/model_runner.py]: Maybe just change `kernel_block_sizes` var above to `self.kernel_block_sizes` instead?

File: vllm/v1/worker/gpu/model_runner.py (hunk 2 above)
Comment (njhill) [inline vllm/v1/worker/gpu/model_runner.py]: I know this is preexisting, but we could move the init_meta logic into the constructor here?

File: vllm/v1/worker/gpu/model_runner.py [hunk 4]
```diff
@@ -446,6 +448,22 @@ def initialize_kv_cache(self, kv_cache_config: KVCacheConfig) -> None:
...
+
+    def _init_kv_zero_meta(self) -> None:
+        """Build KV-block zeroing metadata; invoked from gpu_worker."""
+        self._kv_block_zeroer = KVBlockZeroer(self.device, is_pin_memory_available())
+        self._kv_block_zeroer.init_meta(
+            attn_groups_iter=(g for groups in self.attn_groups for g in groups),
+            kernel_block_sizes=self._kernel_block_sizes,
+            cache_dtype=self.cache_config.cache_dtype,
+            runner_only_attn_layers=set(),
```
Comment (njhill) [inline vllm/v1/worker/gpu/model_runner.py]: make this arg optional?

**Review by njhill — COMMENTED** [review]
Thanks @izhuhaoran

File: vllm/v1/worker/gpu/model_runner.py (hunk 1 above)
Comment (izhuhaoran) [reply vllm/v1/worker/gpu/model_runner.py]: Done. Dropped the wrapper, declared kv_block_zeroer in __init__, and call it directly with an assert.

File: vllm/v1/worker/gpu/model_runner.py (hunk 2 above)
Comment (izhuhaoran) [reply vllm/v1/worker/gpu/model_runner.py]: Done, renamed to kv_block_zeroer

File: vllm/v1/worker/gpu/model_runner.py (hunk 3 above)
Comment (izhuhaoran) [reply vllm/v1/worker/gpu/model_runner.py]: Done, promoted to self.kernel_block_sizes at the unpack and dropped the alias.

File: vllm/v1/worker/gpu/model_runner.py (hunk 4 above)
Comment (izhuhaoran) [reply vllm/v1/worker/gpu/model_runner.py]: Done, defaulted to None (normalized to empty set);

File: vllm/v1/worker/gpu/model_runner.py (hunk 2 above)
Comment (izhuhaoran) [reply vllm/v1/worker/gpu/model_runner.py]: Done — folded init_meta into KVBlockZeroer's constructor and dropped the separate method. As it's shared, I updated the V1 call site too.

**Review by njhill — APPROVED** [review]

**Comment by doctorpangloss** [comment]
Heads-up for anyone carrying this `KVBlockZeroer` on a tree between this PR and #51718.

With overlaid/packed KV layouts (every layer's view carved out of one shared block slab at a byte offset, as in `_get_packed_kv_cache_layout` / cross-layer blocks), `dp = kv.data_ptr()` is not the start of page 0 and `kv.stride(block_dim)` is the whole slab. Zeroing one block stride from that pointer covers slab `b` from the layer's offset to the end **and the first `offset` bytes of slab `b+1`**, which is a live block whenever the free queue hands out the lower id later. On DeepSeek V4 this silently wiped one 64-row indexer cache block per request and made identical requests return different text (needle recall at 512k went from 0/3 to 3/3 with the fix); the corruption depends on allocation order, so it looks like kernel nondeterminism.

#51718 avoids it by spanning `kernel_page_bytes` per address instead of the block stride. For older trees the minimal fix is to zero the whole slab once from the backing base, recovered from the view's storage offset (non-packed layouts have a zero storage offset and are unchanged):

```diff
--- a/vllm/v1/worker/utils.py
+++ b/vllm/v1/worker/utils.py
@@ class KVBlockZeroer:
                 kv = static_forward_context[layer_name].kv_cache
                 if not isinstance(kv, torch.Tensor):
                     continue
-                dp = kv.data_ptr()
+                el = kv.element_size()
+                # Packed layouts carve each layer's view out of a shared
+                # block slab at a byte offset; zero from the slab base so a
+                # block id maps onto its own slab and never the next one.
+                dp = kv.data_ptr() - kv.storage_offset() * el
                 if dp in seen_ptrs:
                     continue
                 seen_ptrs.add(dp)

-                el = kv.element_size()
                 cur_bytes = kv.stride(block_dim) * el
```
